File: ProjectEuler/611/C++/square.cpp
#include <iostream>
#include <time.h>
#include <string>
#include <math.h>
#include <tuple>
#include <queue>
#include <set>

using namespace std;

typedef std::tuple<unsigned long,unsigned long> Coordinates;

unsigned long sumSquare(unsigned long x, unsigned long y){
	return x*x+y*y;
}
bool endColumn(unsigned long x, unsigned long y){
	return (y>=x-1);
}
string expectedStr(int secs){
	secs = abs(secs);
	int days = secs /86400;
	secs %= 86400;
	int hours = secs/3600;
	secs %= 3600;
	int minutes = secs/60;
	secs %= 60;
	if(days > 1){
		return to_string(days) +" days " +to_string(hours)+"h"+to_string(minutes)+"m"+to_string(secs)+"s";
	}else if(hours > 1){
		return to_string(hours)+"h"+to_string(minutes)+"m"+to_string(secs)+"s";
	}else if(minutes >1){
		return to_string(minutes)+"m"+to_string(secs)+"s";
	}else {
		return to_string(secs)+"s";
	}
}



int main(){

	clock_t startT;
	clock_t currentT;

	double execTime = 0;
	string expectedTime = "";

	unsigned long infSquare = 4;
	unsigned long nextSquare = 9;
	queue<Coordinates> notFinishedCol = queue<Coordinates>();
	queue<Coordinates> notFinishedColForNextCateg = queue<Coordinates>();
	set<unsigned long> SetOfSquare = set<unsigned long>();
	unsigned long x = 2;
	unsigned long y = 1;
	unsigned long count = 0;
	unsigned long N = 0;
	unsigned long Max = 1e12;
	bool finishingCateg = true;
	unsigned long stop = floor(sqrt(Max));
	unsigned long square = 0;
	startT = clock();


	while(x < stop+1){
		N++;
		square = sumSquare(x,y);


		// Counting process
		if(square <= Max){
			if(SetOfSquare.find(square) != SetOfSquare.end()){
				SetOfSquare.erase(square);
				count--;
			}else{
				SetOfSquare.insert(square);
				count++;
			}
		}

		//Estimate the time 
		if(floor(log10(N)) == log10(N)){
			currentT = clock();
			execTime =  ((float) (currentT-startT)/(CLOCKS_PER_SEC));
			expectedTime =  expectedStr((int) ( (float) execTime/N*Max));
			cout << "Iteration : " << N << " Count : " << count << " Time : "  << execTime << " Expected total duration : "<< expectedTime << endl;
		}


		// Advance by categories
		if((sumSquare(x,y+1)<=nextSquare) && !(endColumn(x,y))){
			y +=1;
		}
		else{
			if(finishingCateg){
				finishingCateg = false;
				SetOfSquare.clear();
				infSquare = nextSquare;
				nextSquare = (x+2)*(x+2);
				if(!endColumn(x,y)){
					notFinishedColForNextCateg.push(make_tuple(x,y+1));
				}
				notFinishedCol = notFinishedColForNextCateg;
				notFinishedColForNextCateg = queue<Coordinates>();

			}else{
				if(!endColumn(x,y)){
					notFinishedColForNextCateg.push(make_tuple(x,y+1));
				}
			}

			// Next elem
			if(!notFinishedCol.empty()){
					Coordinates next = notFinishedCol.front();
					notFinishedCol.pop();
					x = get<0>(next);
					y = get<1>(next);
			}else {
					finishingCateg = true;
					x++;
					y = 1;
			}


		}

	}
	cout << "Solution: " << count << endl;

	return 0;
}

File: ProjectEuler/611/square.py
import numpy as np
from copy import deepcopy
from collections import deque
import datetime

def sumSquare(x,y):
	return x**2+y**2
def endColumn(x,y):
	return (y>=x-1)
def expectedStr(secs):
	secs = abs(int(secs))
	days, secs = divmod(secs, 86400)
	hours, secs = divmod(secs, 3600)
	minutes,secs = divmod(secs, 60)
	if days > 0:
		return '%d days %dh%dm%ds' % (days,hours,minutes,secs)
	elif hours >0:
		return '%dh%dm%ds' % (hours,minutes,secs)
	elif minutes > 0:
		return '%dm%ds' % (minutes,secs)
	else:
		return '%ds' % (secs)

class Iterator611():
	infSquare = 4
	nextSquare = 9
	notFinishedCol = deque()
	notFinishedColForNextCateg = deque()
	SetOfSquare = set()
	x = 2
	y = 1
	count = 0
	N = 0
	Max = 0
	finishingCateg = True
	startTime = datetime.datetime.now()

	def __init__(self,Max):
		self.stop = np.floor(np.sqrt(Max))
		self.Max = Max
	def __iter__(self):
		return self

	def __next__(self):
		if self.x >= self.stop+1:
			raise StopIteration

		# Count iteration
		self.N = self.N+1
		square = sumSquare(self.x,self.y)

		#Count the square.
		if(square <= self.Max):
			if square in self.SetOfSquare:
				self.SetOfSquare.remove(square); 
				self.count -= 1;
			else:
				self.SetOfSquare.add(square)
				self.count+=1;

		#Estimate the time
		if(np.floor(np.log10(self.N)) == np.log10(self.N)):
			cur_time = datetime.datetime.now()
			exection_time = (cur_time - self.startTime).total_seconds()
			expectedTime = expectedStr(exection_time/self.N *self.Max)
			print("Iteration : ", self.N, "Count : ", self.count, "Time : ", exection_time, " Expected total duration : ", expectedTime)

		#Advance by categories.
		if ((sumSquare(self.x,self.y+1) <= self.nextSquare) and not endColumn(self.x,self.y)):
			self.y+=1		
		else:
			if self.finishingCateg:
				#print(" End of category : ", self.infSquare, " - ", self.nextSquare, " Count : ", self.count)
				self.finishingCateg = False
				self.SetOfSquare = set()
				self.infSquare = self.nextSquare
				self.nextSquare = (self.x+2)**2
				if not endColumn(self.x,self.y):
					self.notFinishedColForNextCateg.append((self.x,self.y+1))
				self.notFinishedCol = deepcopy(self.notFinishedColForNextCateg)
				self.notFinishedColForNextCateg = deque()

				(self.x,self.y) = self.nextElem()
			else:
				if not endColumn(self.x,self.y):
					self.notFinishedColForNextCateg.append((self.x,self.y+1))

				(self.x,self.y) = self.nextElem()

		return (self.count)

	def nextElem(self):
		if self.notFinishedCol:
			return self.notFinishedCol.popleft()
		else:
			self.finishingCateg = True
			return (self.x+1, 1)



